### PR TITLE
Several changes to be compatible with the new imdb layout

### PR DESCRIPTION
**Pull request by M0H4MM4D0UTIS**
Xpath support added in this version
Changes have been made to adapt to the new format.
Also compatible with previous template.

### CODE REVIEW

**Comment by tboothman** [comment]
Thanks for this. I had no intention of doing this by myself as it seemed like too much. Hopefully someone making a start will give some other people a push to contribute too.

There's a few things I'd change here like the xpath being a bit strict and I think detect_new_version_imdb should be removed along with some of the old code to deal with the old layout but I'll merge this so there's something to build on.

File: src/Imdb/Title.php [hunk 1]
```diff
@@ -945,12 +945,17 @@ public function get_episode_details()
...
         if ($this->main_plotoutline == "") {
-            $page = $this->getPage("Title");
-            if (preg_match('!class="summary_text">\s*(.*?)\s*</div>!ims', $page, $match)) {
-                $this->main_plotoutline = trim($match[1]);
-            } elseif ($fallback) {
-                $this->main_plotoutline = $this->storyline();
+            if (isset($this->jsonLD()->description)) {
+                $this->main_plotoutline = htmlspecialchars_decode($this->jsonLD()->description, ENT_QUOTES| ENT_HTML5);
+            } else {
+                $page = $this->getPage("Title");
```
Comment (tboothman) [inline src/Imdb/Title.php:951]: Is `$this->jsonLD()->description` not being there possible? If it's not this else block should be removed

File: src/Imdb/Title.php [hunk 2]
```diff
@@ -985,8 +993,11 @@ private function populatePoster()
         if (isset($this->jsonLD()->image)) {
             $this->main_poster = $this->jsonLD()->image;
         }
-        preg_match('!<img [^>]+title="[^"]+Poster"[^>]+src="([^"]+)"[^>]+/>!ims', $this->getPage("Title"), $match);
-        if (!empty($match[1])) {
+        if (preg_match('!<img [^>]+title="[^"]+Poster"[^>]+src="([^"]+)"[^>]+/>!ims', $this->getPage("Title"), $match)
```
Comment (tboothman) [inline src/Imdb/Title.php:996]: Remove this block - it's for the old layout

File: src/Imdb/Title.php [hunk 3]
```diff
@@ -1621,6 +1632,59 @@ public function cast($short = false)
         if (empty($page)) {
             return array(); // no such page
         }
+        if ($short && $this->detect_new_version_imdb()) {
```
Comment (tboothman) [inline src/Imdb/Title.php:1635]: Remove $this->detect_new_version_imdb()

File: src/Imdb/Title.php [hunk 4]
```diff
@@ -1621,6 +1632,59 @@ public function cast($short = false)
...
+            $xpath = $this->getXpathPage("Title");
+            $nodes = $xpath->query("//section[@data-testid='title-cast']/div[2]/div[@data-testid='title-cast-item']");
```
Comment (tboothman) [inline src/Imdb/Title.php:1637]: This can just be `"//div[@data-testid='title-cast-item']"`

File: src/Imdb/Title.php [hunk 5]
```diff
@@ -1621,6 +1632,59 @@ public function cast($short = false)
...
+                if (empty($dir['name'])) {
+                    continue;
+                }
+                $get_role = $xpath->query(".//a[@data-testid='cast-item-characters-link']/span[1]", $node);
+                if($get_role != null){
+                    $dir["role"] = $get_role->item(0)->nodeValue;
+                }
+
+                $get_img = $xpath->query(".//img[@class='ipc-image']", $node);
+                if($get_img != null && $get_img->item(0)->getAttribute("src") != null){
```
Comment (tboothman) [inline src/Imdb/Title.php:1664]: This doesn't work, I get `Error : Call to a member function getAttribute() on null` when running the test case for short cast

Comment (tboothman) [reply src/Imdb/Title.php:1664]: $get_img can not be false if query is provided with a valid node and xpath, you get an empty DOMNodeList 

https://www.php.net/manual/en/domxpath.query.php
`Returns a DOMNodeList containing all nodes matching the given XPath expression. Any expression which does not return nodes will return an empty DOMNodeList.`

**Comment by duck7000** [comment]
Movie photo does not work properly, i only get have the picture?

**Pull request merged**